### PR TITLE
add service wrapper that emit metrics automatically

### PR DESCRIPTION
**Pull request by yiminc**
add tchannel service wrapper that automatically emit metrics on all service calls.

This will emit metrics like:
PollForActivityTask.cadence-request
RespondDecisionTaskCompleted.cadence-latency
....

### CODE REVIEW

File: common/metrics/service_wrapper.go [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+		return scope
+	}
+	scope = w.scope.SubScope(scopeName)
+	w.childScopes[scopeName] = scope
+	w.mutex.Unlock()
+	return scope
+}
+
+func (w *workflowServiceMetricsWrapper) DeprecateDomain(ctx thrift.Context, request *shared.DeprecateDomainRequest) error {
+	scope := w.getScope("DeprecateDomain")
```
Comment (sivakku) [inline common/metrics/service_wrapper.go]: Can we  make const out of scope name for all these strings easier to look what all being reported.

File: common/metrics/service_wrapper.go [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+	scope.Counter(CadenceRequest).Inc(1)
+	startTime := time.Now()
+	err := w.service.DeprecateDomain(ctx, request)
+	scope.Timer(CadenceLatency).Record(time.Now().Sub(startTime))
+	if err != nil {
```
Comment (sivakku) [inline common/metrics/service_wrapper.go]: We should not count error if the error is EntityNotExistsError, that is a valid response.

File: common/metrics/service_wrapper.go [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+	}
+	scope = w.scope.SubScope(scopeName)
+	w.childScopes[scopeName] = scope
+	w.mutex.Unlock()
+	return scope
+}
+
+func (w *workflowServiceMetricsWrapper) DeprecateDomain(ctx thrift.Context, request *shared.DeprecateDomainRequest) error {
+	scope := w.getScope("DeprecateDomain")
+	scope.Counter(CadenceRequest).Inc(1)
```
Comment (sivakku) [inline common/metrics/service_wrapper.go]: May be wrap these three things in a function that takes scope name  may be a function to execute that wraps cadence call.

File: common/metrics/service_wrapper.go [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+	return err
+}
+
+func (w *workflowServiceMetricsWrapper) DescribeDomain(ctx thrift.Context, request *shared.DescribeDomainRequest) (*shared.DescribeDomainResponse, error) {
+	scope := w.getScope("DescribeDomain")
+	scope.Counter(CadenceRequest).Inc(1)
+	startTime := time.Now()
+	result, err := w.service.DescribeDomain(ctx, request)
+	scope.Timer(CadenceLatency).Record(time.Now().Sub(startTime))
+	if err != nil {
```
Comment (sivakku) [inline common/metrics/service_wrapper.go]: same here for: EntityNotExistsError

File: common/metrics/service_wrapper.go [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+	return result, err
+}
+
+func (w *workflowServiceMetricsWrapper) GetWorkflowExecutionHistory(ctx thrift.Context, request *shared.GetWorkflowExecutionHistoryRequest) (*shared.GetWorkflowExecutionHistoryResponse, error) {
+	scope := w.getScope("GetWorkflowExecutionHistory")
+	scope.Counter(CadenceRequest).Inc(1)
+	startTime := time.Now()
+	result, err := w.service.GetWorkflowExecutionHistory(ctx, request)
+	scope.Timer(CadenceLatency).Record(time.Now().Sub(startTime))
+	if err != nil {
```
Comment (sivakku) [inline common/metrics/service_wrapper.go]: same here: for workflow not found.

File: common/metrics/service_wrapper.go [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+	return err
+}
+
+func (w *workflowServiceMetricsWrapper) RespondActivityTaskCompleted(ctx thrift.Context, request *shared.RespondActivityTaskCompletedRequest) error {
+	scope := w.getScope("RespondActivityTaskCompleted")
+	scope.Counter(CadenceRequest).Inc(1)
+	startTime := time.Now()
+	err := w.service.RespondActivityTaskCompleted(ctx, request)
+	scope.Timer(CadenceLatency).Record(time.Now().Sub(startTime))
+	if err != nil {
```
Comment (sivakku) [inline common/metrics/service_wrapper.go]: same here for: ActivityNotExist for all above activity methods.

File: common/metrics/service_wrapper.go (hunk 3 above)
Comment (yiminc) [inline common/metrics/service_wrapper.go]: I tried that and the result does not look better than this.

File: common/metrics/service_wrapper.go (hunk 2 above)
Comment (yiminc) [inline common/metrics/service_wrapper.go]: Why is EntityNotExistsError count as succeed? It indicate something wrong with the request.

File: common/metrics/service_wrapper.go (hunk 4 above)
Comment (yiminc) [inline common/metrics/service_wrapper.go]: same comments as above. This is not a succeed case.

**Comment by yiminc** [comment]
@sivakku , for this wrapper the main idea is we want to capture a generic info regarding every service call. This is handled in the same way as it is on server side (metricClient).

File: common/metrics/service_wrapper.go (hunk 2 above)
Comment (sivakku) [inline common/metrics/service_wrapper.go]: These counters indicate server failures to respond, but there could be valid response that the domain doesn't exist and the user is retrying crazily on it and that shouldn't count against the failures of the calls. we do this on server side as well, take a look at some of those front-end handler metrics. These failures can cause clients to contact us, but those are valid.

Comment (yiminc) [inline common/metrics/service_wrapper.go]: Those metrics are meant to be used as "tracing level" metrics for us (cadence team). Our users are recommended to monitor based on the higher level metrics that we logged.

**Pull request merged**